### ISSUE TITLE
0.17.0-beta.21 -- Mismatched sources content for...

### ISSUE THREAD

**Issue by joeldenning** [jspm/jspm-cli#1912]
I've encountered this on two projects now -- for one of them going back to 0.17.0-beta.16 worked (and beta.17 didn't). Here is my build script:

```bash
#!/bin/bash -e

rm -rf sofe

./node_modules/jspm/jspm.js build \
	src/engagements-menu.js \
	- lodash \
	- rx \
	- single-spa \
	- canopy-urls!sofe \
	- sentry-error-logging!sofe \
	- toast-service!sofe \
	- fetcher!sofe \
	sofe/engagements-menu.js \
	--format amd \
	--source-map-contents \
	"$@"
```

Here's the error I get:

```
     Creating the single-file build for src/engagements-menu.js - lodash - rx -
     single-spa - canopy-urls!sofe - sentry-error-logging!sofe - toast-service!sofe -
     fetcher!sofe...

err  Error: Mismatched sourcesContent for: src/engagements-menu.js
    at /Users/joel/workplace/engagements-menu/node_modules/systemjs-builder/lib/sourcemaps.js:54:21
    at Array.forEach (native)
    at exports.concatenateSourceMaps (/Users/joel/workplace/engagements-menu/node_modules/systemjs-builder/lib/sourcemaps.js:42:19)
    at createOutput (/Users/joel/workplace/engagements-menu/node_modules/systemjs-builder/lib/output.js:63:21)
    at exports.writeOutputs (/Users/joel/workplace/engagements-menu/node_modules/systemjs-builder/lib/output.js:150:16)
    at /Users/joel/workplace/engagements-menu/node_modules/systemjs-builder/lib/builder.js:727:14
    at tryCatcher (/Users/joel/workplace/engagements-menu/node_modules/bluebird/js/release/util.js:16:23)
    at Promise._settlePromiseFromHandler (/Users/joel/workplace/engagements-menu/node_modules/bluebird/js/release/promise.js:502:31)
    at Promise._settlePromise (/Users/joel/workplace/engagements-menu/node_modules/bluebird/js/release/promise.js:559:18)
    at Promise._settlePromise0 (/Users/joel/workplace/engagements-menu/node_modules/bluebird/js/release/promise.js:604:10)
    at Promise._settlePromises (/Users/joel/workplace/engagements-menu/node_modules/bluebird/js/release/promise.js:683:18)
    at Async._drainQueue (/Users/joel/workplace/engagements-menu/node_modules/bluebird/js/release/async.js:138:16)
    at Async._drainQueues (/Users/joel/workplace/engagements-menu/node_modules/bluebird/js/release/async.js:148:10)
    at Immediate.Async.drainQueues [as _onImmediate] (/Users/joel/workplace/engagements-menu/node_modules/bluebird/js/release/async.js:17:14)
    at tryOnImmediate (timers.js:534:15)
    at processImmediate [as _immediateCallback] (timers.js:514:5)
```

**Comment by blittle** [comment]
@joeldenning you gotta simplify that arithmetic build ;) 

```
#!/bin/bash -e

rm -rf sofe

./node_modules/jspm/jspm.js build \
    src/engagements-menu.js \
    - lodash \
    - rx \
    - single-spa \
    - *!sofe \
    sofe/engagements-menu.js \
    --format amd \
    --source-map-contents \
    "$@"
```

**Comment by joeldenning** [comment]
My workaround for now is to either go back to beta.16 or to `--skip-source-maps`

**Comment by joeldenning** [comment]
I think I've found what might cause this problem. I have a file that looks like this:

```js
import React from 'react';

export default function Loader(props) {
	return (
		<div>
			{doneWaiting(props.asyncChecks, props)
				? props.children
				: <div className={`cps-loader`}>
						<span></span>
						<span></span>
						<span></span>
					</div>
			}
		</div>
	)
}

function doneWaiting() {
	return asyncChecks.reduce((bool, func) => bool && func(props), true);
}

export { doneWaiting };
```

It seems like trying to export a function that is declared in a module is related to the problem. I am using plugin-babel for transpilation. The way I know that it is related to the `export { doneWaiting };` is because if I change to the following there is no error.

```js
import React from 'react';

export default function Loader(props) {
	return (
		<div>
			{_doneWaiting(props.asyncChecks, props)
				? props.children
				: <div className={`cps-loader`}>
						<span></span>
						<span></span>
						<span></span>
					</div>
			}
		</div>
	)
}

export function doneWaiting() {
	return _doneWaiting.apply(this, arguments);
}

function _doneWaiting(asyncChecks, props) {
	return asyncChecks.reduce((bool, func) => bool && func(props), true);
}
```

**Comment by joeldenning** [comment]
Actually it's now happening to that same file even when it does not have that type of export:

```js
import React from 'react';

export default function Loader(props) {
	return (
		<div>
			{_doneWaiting(props.asyncChecks, props)
				? props.children
				: <div className={`cps-loader`}>
						<span></span>
						<span></span>
						<span></span>
					</div>
			}
		</div>
	)
}

export function doneWaiting() {
	return _doneWaiting.apply(this, arguments);
}

function _doneWaiting(asyncChecks, props) {
	const result = asyncChecks.reduce((bool, func) => bool && func(props), true);
	console.log('result = ' + result)
	return result;
}
```

**Comment by joeldenning** [comment]
Apologize for the spam, but my new theory is that it's actually related to the build watcher. It seems to work the first time but then get messed up on subsequent watcher runs.

**Comment by guybedford** [comment]
There may be some kind of mutation of the `load.metadata.sourceMap` happening here. Any mutation here should be detectable through debugging the build tree load record for that `src/engagements-menu.js` file before and after the first build.

**Comment by joeldenning** [comment]
@guybedford I have experienced this problem intermittently but when it happened again today I decided to make a repo that reproduces this problem so that you could take a look at it. I can reproduce this problem on beta.28 with only ~10 lines of code. It seems like somehow it might be related to exporting objects that use an imported value as a property in the object. Or at least that's what it seems to be.

When I add some console log statements into the [relevant systemjs-builder code](https://github.com/systemjs/builder/blob/master/lib/sourcemaps.js#L54), the problem seems to be that `contentsBySource` has the content for the wrong file.

Here were my log statements:
<img width="675" alt="screen shot 2016-09-26 at 10 21 55 pm" src="https://cloud.githubusercontent.com/assets/5524384/18860128/a9802940-8437-11e6-8b59-9c64e5d5ec1f.png">

And here is what was logged:
<img width="740" alt="screen shot 2016-09-26 at 10 22 27 pm" src="https://cloud.githubusercontent.com/assets/5524384/18860131/bd725ca2-8437-11e6-996b-38006afc3036.png">

Note that in the above screenshot that the content that is supposedly for the [mismatched-source-contents.js file](https://github.com/joeldenning/mismatched-sources-content/blob/master/src/mismatched-sources-content.js) but that it is actually the content for [root.component.js](https://github.com/joeldenning/mismatched-sources-content/blob/master/src/root.component.js)

The repo that reproduces this problem:
https://github.com/joeldenning/mismatched-sources-content

**Comment by guybedford** [comment]
@joeldenning would you be interested in working on a PR here at all?

**Comment by joeldenning** [comment]
I'll give it a shot -- if it turns out to be an issue with how babel is creating sourcemaps then that might be tough for me to pin down. But assuming that it's just systemjs-builder that is mixing up which source file goes to which source map, I think I could probably fix it.

**Comment by guybedford** [comment]
@joeldenning that would be amazing! @asapach has also been working on the source maps in https://github.com/asapach/builder/tree/sergei-startsev-sourcemaps, I don't know if it will be related though.

**Comment by joeldenning** [comment]
I found the code to change to fix this problem, only to see that @laurentgoudet got to it first and fixed it nine days ago with https://github.com/systemjs/builder/pull/715. Also note that this issue is a duplicate of https://github.com/systemjs/builder/issues/678.

The problem is that the workaround [here](https://github.com/systemjs/builder/blob/d20a7f772fa8c44f257cc925ab217488fd25d336/compilers/esm.js#L122) (for when babel doesn't respect the sourceFileName) doesn't work when the source map has multiple `sources` and `sourcesContent`, which is what happens when rollup inlines modules into other modules. The workaround assumed that there was only one item in the array or that the first item in the `sources` array had been messed up by babel. But that is not the case.

Closing this, since this will be released whenever a new version of systemjs-builder is released. Thanks @laurentgoudet for fixing this!
